swirl_lm/boundary_condition/immersed_boundary_method.py

```diff
@@ -317,6 +328,507 @@ def _check_mesh_size(a, b):
+      image point."""
+
+      node_dist = []
+      temp_mask = []
+
+      for i, j, k in itertools.product(range(2), range(2), range(2)):
+        p_x = common_ops.gather(xx,
+            tf.stack([
+                lower_left_index[0] + i,
+                lower_left_index[1] + j,
+                lower_left_index[2] + k
+            ], axis=1),
+        )
+        p_y = common_ops.gather(yy,
+            tf.stack([
+                lower_left_index[0] + i,
+                lower_left_index[1] + j,
+                lower_left_index[2] + k
+            ], axis=1),
+        )
+        p_z = common_ops.gather(zz,
+            tf.stack([
+                lower_left_index[0] + i,
+                lower_left_index[1] + j,
+                lower_left_index[2] + k
+            ], axis=1),
+        )
+
+        node_dist.append(
+            tf.math.sqrt(
+                (image_point[0] - p_x) ** 2 +
+                (image_point[1] - p_y) ** 2 +
+                (image_point[2] - p_z) ** 2))
+
+        # Flag corner points with 1 which are in the fluid domain, and
+        # set all corner points to zero in the solid domain
+        temp_mask.append(
+            common_ops.gather(tf.transpose(ib_interior_mask, perm=[1, 2, 0]),
+                tf.stack([lower_left_index[0] + i,
+                          lower_left_index[1] + j,
+                          lower_left_index[2] + k],
+                axis=1))
+        )
+
+      fluid_corner_point_mask = tf.nest.map_structure(
+          lambda _v: tf.where(_v < 0.75, tf.zeros_like(_v), tf.ones_like(_v)),
+          temp_mask,
+      )
+
+      # We mask out all enclosing grid nodes that lay in the solid domain by
+      # multiplying their distance to the image point with zero, while all grid
+      # nodes in the solid are masked with 1 (i.e., our
+      # `fluid_corner_point_mask`).
+      node_dist = tf.nest.map_structure(tf.math.multiply, node_dist,
+                                        fluid_corner_point_mask)
+
+      # Flag all image points that have less than 8 enclosing grid nodes
+      # in the FLUID domain. If all enclosing grid nodes for a given image point
+      # are inside the fluid domain, the sum of `fluid_corner_point_mask` will be
+      # exactly 8 (since there are 8 enclosing nodes in 3D space for an image
+      # point). However, if one or more grid nodes are inside the solid domain,
+      # the corresponding sum of `fluid_corner_point_mask` will be less than 8,
+      # which indicates that we have a degenerated cell. Note that
+      # `fluid_corner_point_mask` already flags every node with `1` if inside the
+      # fluid domain, and `0` if inside the solid domain.
+      # Source: Iaccarino and Verzicco, Applied Mechanics Reviews 56(3), 2003.
+      degenerated_cell_mask = tf.nest.map_structure(
+          lambda v: tf.where(v < 7.5, tf.ones_like(v), tf.zeros_like(v)),
+          tf.math.reduce_max(fluid_corner_point_mask, axis=0)
```

Comment:
This should be reduce_sum❗

v = tf.math.reduce_sum(tf.stack(fluid_corner_point_mask))
degenerated_cell_mask = tf.where(v < 7.5, tf.ones_like(v), tf.zeros_like(v))